File: pieces/knight.py
from __future__ import annotations

from pieces.piece import Piece, BoardGrid
from core.cell import Cell
from core.constants import PieceValue

_KNIGHT_DELTAS = [(-2, -1), (-2, 1), (-1, -2), (-1, 2), (1, -2), (1, 2), (2, -1), (2, 1)]
"""All eight (dr, dc) offsets for a knight's L-shaped jump."""


class Knight(Piece):
    """A knight chess piece that jumps in an L-shape.

    Unlike sliding pieces, the knight leaps over any intervening pieces.
    A pinned knight (self.critical == True) will have all its moves filtered
    away by filter_to_pin_ray(), since it cannot move along a pin ray.
    """

    def calc_targets(self, board: BoardGrid) -> None:
        """Populates self.targets with all squares reachable by this knight.

        Checks all eight L-shaped jump destinations. A square is reachable if
        it is on the board and not occupied by a friendly piece. Applies pin
        filtering if the knight is pinned (self.critical is True), which in
        practice removes all targets since a knight cannot move along a ray.

        Args:
            board: The current 8x8 board state.
        """
        self.targets = []
        for dr, dc in _KNIGHT_DELTAS:
            r, c = self.row + dr, self.col + dc
            if 0 <= r <= 7 and 0 <= c <= 7:
                occupant = board[r][c]
                if occupant is None or occupant.team != self.team:
                    self.targets.append(Cell(r, c))
        if self.critical:
            super().critical_man()

    def get_value(self, board: BoardGrid) -> int:
        """Returns the heuristic value of this knight.

        Base value from PieceValue.KNIGHT, plus one point per reachable square,
        and an extra point for each target in the four central squares
        (rows 3–4, cols 3–4).

        Args:
            board: The current 8x8 board state.

        Returns:
            Integer heuristic score for this knight.
        """
        total = PieceValue.KNIGHT
        self.calc_targets(board)
        for cell in self.targets:
            total += 1
            if (cell.row == 3 or cell.row == 4) and (cell.col == 3 or cell.col == 4):
                total += 1
        return total

    def print_piece(self) -> None:
        """Prints the piece type and current position to stdout."""
        print("Knight at", self.row, ",", self.col)

File: Chess101iOS/Sources/Chess101Engine/Knight.swift
public final class Knight: Piece {
    public var row: Int
    public var col: Int
    public var team: Team
    public var targets: [Cell] = []
    public var touched: Bool = false
    public var critical: Bool = false
    public var criticalTargets: [Cell] = []
    public var typeName: String { "Knight" }

    private static let moves = [(-2,-1),(-2,1),(-1,-2),(-1,2),(1,-2),(1,2),(2,-1),(2,1)]

    public init(row: Int, col: Int, team: Team) {
        self.row = row; self.col = col; self.team = team
    }

    @discardableResult
    public func calcTargets(board: Board) -> Bool? {
        targets = []
        for (dr, dc) in Knight.moves {
            let nr = row + dr, nc = col + dc
            guard nr >= 0 && nr < 8 && nc >= 0 && nc < 8 else { continue }
            if board.grid[nr][nc] == nil || board.grid[nr][nc]!.team != team {
                targets.append(Cell(nr, nc))
            }
        }
        if critical { criticalMan() }
        return nil
    }

    public func move(toRow: Int, toCol: Int, board: Board) -> MoveResult {
        row = toRow; col = toCol; touched = true
        return .normal
    }

    public func getValue(board: Board) -> Int {
        var val = 13
        val += targets.count
        let cr = abs(row - 3) <= 1 ? 1 : 0; let cc = abs(col - 3) <= 1 ? 1 : 0
        val += cr + cc
        return val
    }
}
